### PR TITLE
Implement support for send grid basic auth

### PR DESCRIPTION
**Pull request by ghost**
* Add support for SENDGRID_USER & SENDGRID_PASSWORD => basic auth

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,15 +25,20 @@ require "carbon_sendgrid_adapter"
...
 
-private def send_grid_key_from_env
-  ENV["SEND_GRID_KEY"]? || raise_missing_key_message
+private def construct_send_grid_adapter
+  if ENV["SEND_GRID_KEY"]?
+    Carbon::SendGridAdapter.new(api_key: ENV["SEND_GRID_KEY"])
+  elsif ENV["SENDGRID_USER"]? && ENV["SENDGRID_PASSWORD"]?
+    Carbon::SendGridAdapter.new(username: ENV["SENDGRID_USER"], password: ENV["SENDGRID_PASSWORD"])
+  else
+    raise_missing_key_message
```
Comment (ghost) [inline README.md:40]: Currently only mentions `SEND_GRID_KEY` which is okay I guess? API_KEY should be the default

File: src/carbon_sendgrid_adapter.cr [hunk 2]
```diff
@@ -2,15 +2,37 @@ require "http"
...
+  def initialize(username : String, password : String, @sandbox = false)
+    @authentication_strategy = AuthenticationStrategy.new(username, password)
   end
 
   def deliver_now(email : Carbon::Email)
-    Carbon::SendGridAdapter::Email.new(email, api_key, sandbox?).deliver
+    Carbon::SendGridAdapter::Email.new(email, authentication_strategy, sandbox?).deliver
+  end
+
+  class AuthenticationStrategy
```
Comment (ghost) [inline src/carbon_sendgrid_adapter.cr:21]: A wrapper for api_key and (username / password) :D

File: src/carbon_sendgrid_adapter.cr [hunk 3]
```diff
@@ -2,15 +2,37 @@ require "http"
...
+    private getter password : String?
+
+    def initialize(@username, @password)
+    end
+
+    def initialize(@api_key)
+    end
+    
+    def authenticate_with_basic_auth?
+      api_key.nil?
```
Comment (ghost) [inline src/carbon_sendgrid_adapter.cr:33]: If no error was thrown during the construction of the adapter then a nil api_key means that _USER and _PASSWORD were present and therefore a basic auth is in order

File: src/carbon_sendgrid_adapter.cr [hunk 4]
```diff
@@ -3,14 +3,35 @@ require "json"
 require "carbon"
 
 class Carbon::SendGridAdapter < Carbon::Adapter
-  private getter api_key : String
+  getter authentication_strategy : AuthenticationStrategy
   private getter? sandbox : Bool
 
-  def initialize(@api_key, @sandbox = false)
+  def initialize(api_key : String, @sandbox = false)
```
Comment (ghost) [inline src/carbon_sendgrid_adapter.cr:9]: I'm a big fan off method overloading :D

File: src/carbon_sendgrid_adapter.cr [hunk 5]
```diff
@@ -126,7 +147,11 @@ class Carbon::SendGridAdapter < Carbon::Adapter
     private def client : HTTP::Client
       @_client ||= HTTP::Client.new(BASE_URI, port: 443, tls: true).tap do |client|
         client.before_request do |request|
-          request.headers["Authorization"] = "Bearer #{api_key}"
+          if authentication_strategy.authenticate_with_basic_auth?
+            client.basic_auth(authentication_strategy.username, authentication_strategy.password)
```
Comment (ghost) [inline src/carbon_sendgrid_adapter.cr:151]: https://docs.sendgrid.com/for-developers/sending-email/authentication#basic-authentication

**Review by jwoertink — COMMENTED** [review]
First off, thanks for tackling this. The code looks fine; however... after actually reading that link you pasted, I saw this:

> To ensure the security of your account, Twilio SendGrid has ended support for basic authentication with username and password as of Q4 2020.

It also mentions this:

> When using Basic Authentication, your username will always be "apikey"

Which makes me wonder, should we even support this? I know that the original issue was about how Heroku set things up, but I'm thinking maybe the Issue was posted well before this change was made?

**Comment by ghost** [comment]
@jwoertink agreed 😆 I guess this was a valid strategy in 2018 :D should not be supported now 👍🏻... should have read the page first :D

**Pull request closed without merging**